preprocessing/preprocess.py

```diff
@@ -0,0 +1,157 @@
+
+    negatives_blob = bucket.get_blob('metadata/negatives.csv')
+    negatives_bytes = negatives_blob.download_as_string()
+    negatives_str = negatives_bytes.decode('utf-8')
+    negatives_df = pd.read_csv(io.StringIO(negatives_str))
+    return positives_df, negatives_df
+
+
+def process_cab(blob: storage.Blob, patient_id: str) -> None:
+    os.mkdir('tmp')
+    os.chdir('tmp')
+
+    blob.download_to_filename(patient_id + '.cab')
+    subprocess.call(['cabextract', patient_id + '.cab'])
+    logging.info('extracted cab file')
+
+    dirpath = list(os.walk('.'))[2][0]
+    logging.info(f'loading scans from {dirpath}')
+    processed_scan = _process_cab(dirpath)
+
+    os.chdir('..')
+    shutil.rmtree('tmp')
+
+    save_to_gcs(processed_scan, patient_id, blob.bucket)
+
+
+def _process_cab(dirpath: str) -> np.array:
+    scan = parsers.load_scan(dirpath)
+    processed_scan = preprocess_scan(scan)
+    return processed_scan
+
+
+def process_zip(blob: storage.Blob, patient_id: str) -> None:
+    os.mkdir('tmp')
+    os.chdir('tmp')
+
+    blob.download_to_filename(patient_id + '.zip')
+    logging.info('extracting zip file')
+    shutil.unpack_archive(patient_id + '.zip', format='zip')
+
+    dirpath = list(os.walk('.'))[3][0]
+    logging.info(f'loading scans from {dirpath}')
+    scan = parsers.load_scan(dirpath)
+    processed_scan = preprocess_scan(scan)
+    logging.info(f'processing dicom data')
+
+    os.chdir('..')
+    shutil.rmtree('tmp')
+    save_to_gcs(processed_scan, patient_id, blob.bucket)
+
+
+def save_to_gcs(processed_scan, patient_id, bucket):
+    outpath = f'numpy/{patient_id}.npy'
+    stream = io.BytesIO()
+    np.save(stream, processed_scan)
+    processed_blob = storage.Blob(outpath, bucket=bucket)
+    stream.seek(0)
+    processed_blob.upload_from_file(stream)
+    logging.info(f'saving dicom data to GCS in {outpath}')
+
+
+def create_labels_csv(input_bucket, positives_df, negatives_df) -> None:
+    """Creates a file labels.csv mapping patient_id to 1, if positive
+    and 0, if negative.
+    """
+    labels = []
+    blob: storage.Blob
+    for blob in input_bucket.list_blobs(prefix=ELVOS_ANON + '/'):
+        if blob.name.endswith('.csv'):
+            continue  # Ignore the metadata CSV
+
+        patient_id = blob.name[len(ELVOS_ANON) + 1: -EXTENSION_LENGTH]
+
+        if patient_id in positives_df['Anon ID'].values:
+            labels.append((patient_id, 1))
+        elif patient_id in negatives_df['Anon ID'].values:
+            labels.append((patient_id, 0))
+        else:
+            logging.warning(f'blob with patient id {patient_id} not found in'
+                            f' the metadata CSVs, defaulting to negative')
+            labels.append((patient_id, 0))
+    labels_df = pd.DataFrame(labels, columns=['patient_id', 'label'])
+    labels_df.to_csv('labels.csv', index=False)
+    logging.info(f'label value counts {labels_df["label"].value_counts()}')
```

Comment:
Should we upload the labels.csv file to GCS?

Reply:
Yeah this should be uploaded as well.